pyomo/common/tee.py

```diff
@@ -211,100 +218,170 @@ class capture_output(object):
+                    fd_redirect[i + 1] = self._enter_context(
+                        redirect_fd(i + 1, tee_fd[i], synchronize=False)
+                    )
+                    # Redirect the file descriptor currently associated with
+                    # sys.stdout / sys.stderr
+                    fd = old_fd[i]
+                    if fd and fd not in fd_redirect:
+                        fd_redirect[fd] = self._enter_context(
                             redirect_fd(fd, tee_fd[i], synchronize=False)
                         )
-            for fdr in self.fd_redirect:
-                fdr.__enter__()
-        # We have an issue where we are (very aggressively)
-        # commandeering the terminal.  This is what we intend, but the
-        # side effect is that any errors generated by this module (e.g.,
-        # because the user gave us an invalid output stream) get
-        # completely suppressed.  So, we will make an exception to the
-        # output that we are catching and let messages logged to THIS
-        # logger to still be emitted.
-        if self.capture_fd:
-            # Because we are also comandeering the FD that underlies
-            # self.old[1], we cannot just write to that stream and
-            # instead open a new stream to the original FD.
-            #
-            # Note that we need to duplicate the FD from the redirector,
-            # as it will close the (temporary) `original_fd` descriptor
-            # when it restores the actual original descriptor
-            self.temp_log_stream = os.fdopen(
-                os.dup(self.fd_redirect[-1].original_fd), mode="w", closefd=True
-            )
-        else:
-            self.temp_log_stream = self.old[1]
-        self.temp_log_handler = logging.StreamHandler(self.temp_log_stream)
-        logger.addHandler(self.temp_log_handler)
-        self._propagate = logger.propagate
-        logger.propagate = False
-        return self.output_stream
+            # We need to make sure that we didn't just capture the FD
+            # that underlies a stream that we are outputting to.  Note
+            # that when capture_fd==False, normal streams will be left
+            # alone, but the lastResort _StderrHandler() will still be
+            # replaced (needed because that handler uses the *current*
+            # value of sys.stderr)
+            ostreams = []
+            for stream in self.tee.ostreams:
+                if isinstance(stream, LogStream):
+                    for handler_redirect in stream.redirect_streams(fd_redirect):
+                        self._enter_context(handler_redirect, prior_to=self.tee)
+                else:
+                    try:
+                        fd = stream.fileno()
+                    except AttributeError:
+                        fd = None
+                        # if not hasattr(stream, 'write'):
+                        #    raise
```

Comment:
Is this debugging code that should be removed or history that could use context? If it can be removed then you could combine this `except` with the one below for `OSError`.

Reply:
I think it can be removed.  At one point I was using that to better ensure that the stream was an actual stream (I think I was accidentally getting stream to be a tuple of streams, but that is handled elsewhere now.